Clean garbage tmpdirs with tempfile.TemporaryDirectory

graingert opened the pull request:
Closes #7940

see also #7933 for my first attempt

graingert commented on src/_pytest/pathlib.py:
```diff
@@ -254,11 +254,10 @@ def maybe_delete_a_numbered_dir(path: Path) -> None:
-        garbage = parent.joinpath(f"garbage-{uuid.uuid4()}")
-        path.rename(garbage)
-        rm_rf(garbage)
+        with tempfile.TemporaryDirectory(
```
hmm I think this needs to be 0o755

https://github.com/python/cpython/blob/3d43f1dce3e9aaded38f9a2c73e3c66acf85505c/Lib/tempfile.py#L793

graingert replied:
https://bugs.python.org/issue26660

graingert replied:
ok I'm just doing it wrong - it works here:

```python
import tempfile
import pathlib


def main():
    with tempfile.TemporaryDirectory() as tmp:
        tmp_path = pathlib.Path(tmp)
        p = tmp_path / "ham" / "spam" / "eggs"
        p.mkdir(parents=True)
        p.touch(mode=0)
        for parent in p.parents:
            if parent == tmp_path:
                break
            parent.chmod(mode=0)

    assert not tmp_path.exists()


if __name__ == "__main__":
    main()
```

RonnyPfannschmidt commented on src/_pytest/pathlib.py:
```diff
@@ -254,11 +254,10 @@ def maybe_delete_a_numbered_dir(path: Path) -> None:
-        garbage = parent.joinpath(f"garbage-{uuid.uuid4()}")
-        path.rename(garbage)
-        rm_rf(garbage)
+        with tempfile.TemporaryDirectory(
```
This should spec a root directory and a prefix

graingert replied:
dir is the root directory

bluetech commented:
IIUC, this is basically a hack to coerce `TemporaryDirectory.cleanup()` to do the dirty work for us, right? I like it :) Although I think a comment is in order to explain what it's doing, it seems odd if you don't get it.

bluetech commented on src/_pytest/pathlib.py:
```diff
@@ -254,11 +254,10 @@ def maybe_delete_a_numbered_dir(path: Path) -> None:
-        path.rename(garbage)
-        rm_rf(garbage)
+        with tempfile.TemporaryDirectory(
+            prefix="garbage-", dir=os.fsdecode(path.parent)
```
Why is the `os.fsdecode` needed?

graingert replied:
TemporaryDirectory needs arguements of the same type

graingert replied:
actually dir seems to only be used in `os.path.join` so it's safe as a pathlib object

RonnyPfannschmidt commented:
Since the rmtree helper is simple, how about replicating

graingert commented:
so it turns out this PR does fix my original issue, but my test case is hitting https://github.com/pytest-dev/pytest/issues/7947

graingert committed:
use tempfile.TemporaryDirectory to cleanup garbage- dir

Python has a very nice rm_rf hidden in
tempfile.TemporaryDirectory._rmtree we can use it to cleanup our
"garbage-" dir
https://github.com/python/cpython/blob/3d43f1dce3e9aaded38f9a2c73e3c66acf85505c/Lib/tempfile.py#L784-L812

graingert commented:
> IIUC, this is basically a hack to coerce `TemporaryDirectory.cleanup()` to do the dirty work for us, right? I like it :) Although I think a comment is in order to explain what it's doing, it seems odd if you don't get it.

I'm just composing a comment for this now - and I'd like to comment the use of the rename to "garbage-",  but I'm not really clear on why this rename was done originally: https://github.com/pytest-dev/pytest/pull/3988/files#diff-b2559484fd51ba4a312585ab8375eda68c43b0b66bb329ec036106255928a2b9R169

graingert commented:
this breaks on windows and I don't know why

RonnyPfannschmidt commented:
@graingert the name garbage was replicated from pylib which uses it as indicator of folder names that can be removed even if things fail from the processs that initially tries to take them out

RonnyPfannschmidt commented:
https://github.com/pytest-dev/py/blob/d2a654131d4af0d2c448c56b599e46a2e0725de4/py/_path/local.py#L840-L994
